compiler/executable/update/executable.rs

```diff
@@ -41,42 +45,122 @@ impl UpdateExecutable {
 }
 
 pub fn compile(
-    constraints: &[Constraint<Variable>],
+    block: &Block,
     input_variables: &HashMap<Variable, VariablePosition>,
-    type_annotations: &TypeAnnotations,
+    block_annotations: &BlockAnnotations,
     variable_registry: &VariableRegistry,
     source_span: Option<Span>,
 ) -> Result<UpdateExecutable, Box<WriteCompilationError>> {
     let mut variable_positions = input_variables.clone();
     let attributes_inserts = add_inserted_concepts(
-        constraints,
-        type_annotations,
+        block.conjunction(),
+        block_annotations,
         variable_registry,
         input_variables,
         &mut variable_positions,
         source_span,
     )?;
 
-    let mut connection_inserts = Vec::with_capacity(constraints.len());
-
-    add_has(constraints, &variable_positions, variable_registry, &mut connection_inserts)?;
-    add_links(
-        constraints,
-        type_annotations,
-        input_variables,
+    let connection_instructions = add_connections(
+        block.conjunction(),
+        block_annotations,
         &variable_positions,
         variable_registry,
-        &mut connection_inserts,
+        input_variables,
     )?;
 
+    let mut optional_updates = Vec::with_capacity(block.conjunction().nested_patterns().len());
+    for nested_pattern in block.conjunction().nested_patterns() {
+        let NestedPattern::Optional(optional) = nested_pattern else {
+            unreachable!("Only optionals are allowed as nested patterns in update")
```

Comment:
Or rather - 
Non-optional update patterns are illegal and should have been rejected earlier

?